core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1170,6 +1174,8 @@ class KafkaApis(val requestChannel: RequestChannel,
           debug(s"OffsetRequest with correlation id $correlationId from client $clientId on partition $topicPartition " +
               s"failed because the partition is duplicated in the request.")
           buildErrorResponse(Errors.INVALID_REQUEST, partition)
+        } else if (timestampMinSupportedVersion.contains(partition.timestamp()) && version < timestampMinSupportedVersion(partition.timestamp())) {
```

Comment:
Personally, I love this strict check to avoid "undefined" behavior. However, I also try to raise "possible" objection. 

**Should we view "negative timestamp" as user-defined data?** 

I had saw users who using the negative timestamp (yes, we don't disable that in protocol) in their data. That means the "negative timestamp" could be either user-defined data or flag. Hence, those new flag (and version check) can break the behavior that tries to find the offset of user-defined negative timestamp. We can fix it by adding a new field `timestampType` to `LIST_OFFSET` request. The new field can be used to indicate the "type" of timestamp.

Reply:
for another one: 

**Should we keep the behavior of timestamp and unmatched version?**

the older broker which does not know either `EARLIEST_LOCAL_TIMESTAMP` or `LATEST_TIERED_TIMESTAMP` will process the timestamp "flag" as normal timestamp (do the search). That means the new broker will return "different" result - that seems to be a kind of behavior broken - the request having old version should be treated as "old" behavior

Reply:
It seems we try to keep the behavior only on version_0 [0]. After that, we just add more `if-else` to handle each timestamp flag without handling above issues. If that is the expected changes, maybe the jira is overkill, and we can close this PR to simplify broker's behavior.

[0] https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/KafkaApis.scala#L1078


@FrankYang0529 @junrao WDYT?

Reply:
Hi @chia7712, thanks for the great questions! First, we should have consensus about "how to handle a new defined behavior which was undefined in before"? So we can know how to answer further questions.

IIRC, for negative timestamp, we don't support it in Kafka. I found [KIP-228](https://cwiki.apache.org/confluence/display/KAFKA/KIP-228%2BNegative%2Brecord%2Btimestamp%2Bsupport) about it, but it looks like it's still under discussion. IMHO, if a behavior is unexpected, we may need to add some limitations to it in a newer version, so we can minimize unexpected result.

Reply:
> IIRC, for negative timestamp, we don't support it in Kafka. I found [KIP-228](https://cwiki.apache.org/confluence/display/KAFKA/KIP-228%2BNegative%2Brecord%2Btimestamp%2Bsupport) about it, but it looks like it's still under discussion. IMHO, if a behavior is unexpected, we may need to add some limitations to it in a newer version, so we can minimize unexpected result.

thanks for that sharing. That means we do NOT care the cases of using negative ts in record. Hence, we can add explicit behavior for each version, and "reject" the requests which using "unknown" negative ts. for example:
```scala
else if (partition.timestamp() < 0 && !timestampMinSupportedVersion.contains(partition.timestamp())) {
          buildErrorResponse(Errors.UNSUPPORTED_VERSION, partition)
        } 
```

@FrankYang0529 WDYT?

Reply:
Hi @chia7712 and @FrankYang0529 , based on our discussion, I’ve summarized the current handling of specific negative timestamps in Kafka. Here are the mappings (`timestamp : version`)
	•	`LATEST_TIMESTAMP (-1L) : 1`
	•	`EARLIEST_TIMESTAMP (-2L) : 1`
	•	`MAX_TIMESTAMP (-3L) : 7`
	•	`EARLIEST_LOCAL_TIMESTAMP (-4L) : 8`
	•	`LATEST_TIERED_TIMESTAMP (-5L) : 9`

This should provide a clear guideline for the expected behavior across different versions when dealing with these negative timestamps.

Reply:
Hi @FrankYang0529 and @chia7712,

In the ListOffsetsRequest, we have the forConsumer field that was recently updated to handle different timestamps. Perhaps we could define the behavior based on this logic.

```java
        public static Builder forConsumer(boolean requireTimestamp,
                                          IsolationLevel isolationLevel,
                                          boolean requireMaxTimestamp,
                                          boolean requireEarliestLocalTimestamp,
                                          boolean requireTieredStorageTimestamp) {
            short minVersion = 0;
            if (requireTieredStorageTimestamp)
                minVersion = 9;
            else if (requireEarliestLocalTimestamp)
                minVersion = 8;
            else if (requireMaxTimestamp)
                minVersion = 7;
            else if (isolationLevel == IsolationLevel.READ_COMMITTED)
                minVersion = 2;
            else if (requireTimestamp)
                minVersion = 1;
            return new Builder(minVersion, ApiKeys.LIST_OFFSETS.latestVersion(), CONSUMER_REPLICA_ID, isolationLevel);
        }
```

Reply:
Hi @chia7712 and @frankvicky, thanks for the information. Updated it. Could you take a look again when you have time? Thank you.